### PR TITLE
Added signature support for ECC

### PR DESCRIPTION
**Pull request by sergiizubchevskyy**

### CODE REVIEW

File: src/mococrw/signature.h [hunk 1]
```diff
@@ -33,6 +33,35 @@ class SignatureUtils
 public:
     SignatureUtils() = delete;
 
+    /**
+     * @brief Sign a message with ECC signature
+     *
+     * Creates an ECC signature for a plain text message based on a given private key.
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/mococrw/signature.h]: [Sign vs. Creates]:
In the @brief section it is stated "Sign a message". Here we state that "Creates an ECC signature". IMO this is a little bit confusing. This method signs a message and returns a signed message? Or returns the signature for the given message?

File: src/mococrw/signature.h [hunk 2]
```diff
@@ -33,6 +33,35 @@ class SignatureUtils
...
+     *
+     * Creates an ECC signature for a message digest based on a given private key.
+     *
+     * @param privateKey The private key to be used in the signature
+     * @param digestType The digest type
+     * @param messageDigest The message digest
+     * @return The created ECC signature of the message
+     * @throw MoCOCrWException if the sign operation fails.
+     */
+    static std::vector<uint8_t> create(AsymmetricPrivateKey &privateKey,
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/mococrw/signature.h]: In this file, there are a lot of create methods, with distinct signatures. How does a user know that this method creates an ECC signature and not other type of signature? From the usability perspective, I consider this somehow confusing, and imposes that the user of this code needs to have a deep knowledge about cryptography in order to be able to effectively use the code. Please consider to review the concept behind the naming of these methods.

For instance, you can use namespaces to separate ECC, RSA, and so on, functionalities:
auto signedMessage = mococrw::ecc::create(...);
auto signedMessage = mococrw::rsa::create(...);

Or you can rename the methods:
auto signedMessage = mococrw::createECC(...);
auto signedMessage = mococrw::createRSA(...);

File: src/mococrw/signature.h [hunk 3]
```diff
@@ -33,6 +33,35 @@ class SignatureUtils
...
+     *
+     * @param privateKey The private key to be used in the signature
+     * @param digestType The digest type
+     * @param message The message to be hashed and signed
+     * @return The created signature of the message
+     * @throw MoCOCrWException if the sign operation fails.
+     */
+    static std::vector<uint8_t> create(AsymmetricPrivateKey &privateKey,
+                                        const openssl::DigestTypes digestType,
+                                        const std::string message);
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/mococrw/signature.h]: Consider to user const std::string& instead.

File: src/mococrw/signature.h [hunk 4]
```diff
@@ -33,6 +33,35 @@ class SignatureUtils
...
+     *
+     * @param privateKey The private key to be used in the signature
+     * @param digestType The digest type
+     * @param messageDigest The message digest
+     * @return The created ECC signature of the message
+     * @throw MoCOCrWException if the sign operation fails.
+     */
+    static std::vector<uint8_t> create(AsymmetricPrivateKey &privateKey,
+                                       const openssl::DigestTypes digestType,
+                                       const std::vector<uint8_t> messageDigest);
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/mococrw/signature.h]: Consider user const std::vector& instead.

File: src/mococrw/signature.h [hunk 5]
```diff
@@ -95,6 +124,38 @@ class SignatureUtils
...
+     *
+     * @param publicKey The public key to be used in the verification
+     * @param digestType The digest type
+     * @param signature The signature to be verified
+     * @param message The plain text message the signature is verified with
+     * @throw MoCOCrWException if the verification fails.
+     */
+    static void verify(AsymmetricPublicKey &publicKey,
+                       const openssl::DigestTypes digestType,
+                       const std::vector<uint8_t> signature,
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/mococrw/signature.h]: Consider use const std::vector& instead.
Consider user const std::string& instead (next line).

File: src/mococrw/signature.h [hunk 6]
```diff
@@ -95,6 +124,38 @@ class SignatureUtils
...
+     *
+     * @param publicKey The public key to be used in the verification
+     * @param digestType The digest type
+     * @param signature The signature to be verified
+     * @param messageDigest The message digest the signature is verified with
+     * @throw MoCOCrWException if the verification fails.
+     */
+    static void verify(AsymmetricPublicKey &publicKey,
+                       const openssl::DigestTypes digestType,
+                       const std::vector<uint8_t> signature,
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/mococrw/signature.h]: Consider use const std::vector& instead.
Consider user const std::string& instead (next line).

File: src/mococrw/signature.h [hunk 7]
```diff
@@ -129,6 +190,37 @@ class SignatureUtils
...
+     *
+     * @param certificate A x509 certificate from which the public key will be extracted
+     * @param digestType The digest type
+     * @param signature The signature to be verified
+     * @param message The plain text message the signature is verified with
+     * @throw MoCOCrWException if the verification fails.
+     */
+    static void verify(const X509Certificate &certificate,
+                       const openssl::DigestTypes digestType,
+                       const std::vector<uint8_t> signature,
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/mococrw/signature.h]: Consider use const std::vector& instead.
Consider user const std::string& instead (next line).

File: src/mococrw/signature.h [hunk 8]
```diff
@@ -129,6 +190,37 @@ class SignatureUtils
...
+     *
+     * @param certificate A x509 certificate from which the public key will be extracted
+     * @param digestType The digest type
+     * @param signature The signature to be verified
+     * @param messageDigest The message digest the signature is verified with
+     * @throw MoCOCrWException if the verification fails.
+     */
+    static void verify(const X509Certificate &certificate,
+                       const openssl::DigestTypes digestType,
+                       const std::vector<uint8_t> signature,
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/mococrw/signature.h]: Consider use const std::vector& instead. The same applies for the next line.

File: src/signature.cpp [hunk 9]
```diff
@@ -26,6 +26,60 @@ namespace mococrw {
...
+                            const DigestTypes digestType,
+                            const std::vector<uint8_t> messageDigest)
+{
+    std::vector<uint8_t> signature;
+    const auto keyCtx = _EVP_PKEY_CTX_new(privateKey.internal());
+    std::unique_ptr<unsigned char, SSLFree<unsigned char>> sig;
+    size_t siglen;
+
+    try {
+        if (!keyCtx.get()) {
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/signature.cpp:116]: Consider put these 3 lines of code immediately after creating the context (after line 50). This way we create and immediately check the context.

File: src/signature.cpp [hunk 10]
```diff
@@ -26,6 +26,60 @@ namespace mococrw {
...
+    }
+
+    return create(privateKey, digestType, messageDigest);
+}
+
+std::vector<uint8_t> SignatureUtils::create(AsymmetricPrivateKey &privateKey,
+                            const DigestTypes digestType,
+                            const std::vector<uint8_t> messageDigest)
+{
+    std::vector<uint8_t> signature;
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/signature.cpp]: nnot used. Please remove.

File: src/signature.cpp [hunk 11]
```diff
@@ -134,6 +188,39 @@ void SignatureUtils::verify(AsymmetricPublicKey &publicKey,
...
 }
 
+void SignatureUtils::verify(AsymmetricPublicKey &publicKey,
+                            const openssl::DigestTypes digestType,
+                            const std::vector<uint8_t> signature,
+                            const std::string message)
+{
+    std::vector<uint8_t> messageDigest;
+
+    messageDigest = digestMessage(message, digestType);
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/signature.cpp]: In line 36, we added a try-catch around this method. Here we didn't follow the same implementation. Please check if both are correct.

File: src/signature.cpp [hunk 12]
```diff
@@ -134,6 +188,39 @@ void SignatureUtils::verify(AsymmetricPublicKey &publicKey,
...
+    verify(publicKey, digestType, signature, messageDigest);
+}
+
+void SignatureUtils::verify(AsymmetricPublicKey &publicKey,
+                            const openssl::DigestTypes digestType,
+                            const std::vector<uint8_t> signature,
+                            const std::vector<uint8_t> messageDigest)
+{
+    const auto keyCtx = _EVP_PKEY_CTX_new(publicKey.internal());
```
Comment (goncalogouveiacriticaltechworkscom) [inline src/signature.cpp]: Here we are not checking if we got a valid context. Please add the protection.

**Review by goncalogouveiacriticaltechworkscom — CHANGES_REQUESTED** [review]
Please check my comments.
note: as requested, tests not reviewed.

File: src/mococrw/signature.h (hunk 1 above)
Comment (sergiizubchevskyy) [reply src/mococrw/signature.h]: Corrected.

File: src/mococrw/signature.h (hunk 3 above)
Comment (sergiizubchevskyy) [reply src/mococrw/signature.h]: Corrected.

File: src/mococrw/signature.h (hunk 4 above)
Comment (sergiizubchevskyy) [reply src/mococrw/signature.h]: Corrected.

File: src/mococrw/signature.h (hunk 5 above)
Comment (sergiizubchevskyy) [reply src/mococrw/signature.h]: Corrected.

File: src/mococrw/signature.h (hunk 6 above)
Comment (sergiizubchevskyy) [reply src/mococrw/signature.h]: Corrected.

File: src/mococrw/signature.h (hunk 7 above)
Comment (sergiizubchevskyy) [reply src/mococrw/signature.h]: Corrected.

File: src/mococrw/signature.h (hunk 8 above)
Comment (sergiizubchevskyy) [reply src/mococrw/signature.h]: Corrected.

File: src/signature.cpp (hunk 10 above)
Comment (sergiizubchevskyy) [reply src/signature.cpp]: Done.

File: src/signature.cpp (hunk 9 above)
Comment (sergiizubchevskyy) [reply src/signature.cpp:116]: Done.

File: src/signature.cpp (hunk 11 above)
Comment (sergiizubchevskyy) [reply src/signature.cpp]: Harmonized. Added try/catch block

File: src/signature.cpp (hunk 12 above)
Comment (sergiizubchevskyy) [reply src/signature.cpp]: Updated, moved inside "try" block. EVP_PKEY_CTX_new can throw OpenSSL exception if the pointer is null.

File: src/mococrw/signature.h [hunk 13]
```diff
@@ -33,103 +33,10 @@ class SignatureUtils
...
-     * @param padding The padding mode
-     * @param signature The signature to be verified
-     * @param message The plain text message the signature is verified with
-     * @throw MoCOCrWException if the verification fails.
-     */
-    static void verify(const X509Certificate &certificate,
-                       const RSAPadding &padding,
-                       const std::vector<uint8_t> signature,
-                       const std::string message);
+private:
```
Comment (pmvfonseca) [inline src/mococrw/signature.h]: Minor nitpick: I think we should switch the order of the private and public methods. It increases class readability IMO.

File: src/mococrw/signature.h [hunk 14]
```diff
@@ -151,19 +58,237 @@ class SignatureUtils
...
+         * @return The created ECC signature of the message
+         * @throw MoCOCrWException if the sign operation fails.
+         */
+        static std::vector<uint8_t> create(AsymmetricPrivateKey &privateKey,
+                                           const std::vector<uint8_t> &messageDigest);
+
+        /**
+         * @brief Verify a signature (ECC)
+         *
+         * Verifies a signature based on the signed message and a given public key and padding mode.
```
Comment (pmvfonseca) [inline src/mococrw/signature.h]: AFAIK when signing with ECC there is no padding :)

File: src/mococrw/signature.h [hunk 15]
```diff
@@ -151,19 +58,237 @@ class SignatureUtils
...
+         * @throw MoCOCrWException if the verification fails.
+         */
+        static void verify(AsymmetricPublicKey &publicKey,
+                           const std::vector<uint8_t> &signature,
+                           const std::string &message);
+
+        /**
+         * @brief Verify a signature (ECC)
+         *
+         * Verifies a signature based on the message digest and a given public key and padding mode.
```
Comment (pmvfonseca) [inline src/mococrw/signature.h]: See comment above.

File: src/mococrw/signature.h [hunk 16]
```diff
@@ -151,19 +58,237 @@ class SignatureUtils
...
+    /*
+     * Set up an EVP_PKEY_CTX for RSA-based signature and verification operations
+     */
+    static void setUpContext(EVP_PKEY_CTX *ctx,
+                             const RSAPadding &padding,
+                             const OperationType &operation);
 
+    /*
+     * Performs the actual OpenSSL sign operation based on a context and a message digest.
+     * Returns the signed message in @signedMessage
```
Comment (pmvfonseca) [inline src/mococrw/signature.h]: Doesn't it return only the signature itself?

File: src/mococrw/signature.h [hunk 17]
```diff
@@ -151,19 +58,237 @@ class SignatureUtils
...
+     */
+    static void create(EVP_PKEY_CTX *ctx,
+                       const std::vector<uint8_t> &messageDigest,
+                       std::vector<uint8_t> &signedMessage);
+
+    /*
+     * Performs the actual OpenSSL verification operation based on a signature and a message digest.
+     * Propagates the OpenSSLException in case of failure.
```
Comment (pmvfonseca) [inline src/mococrw/signature.h]: Does this comment also apply on the method above?

File: src/signature.cpp [hunk 18]
```diff
@@ -26,9 +26,168 @@ namespace mococrw {
...
 using namespace openssl;
 
-std::vector<uint8_t> SignatureUtils::create(AsymmetricPrivateKey &privateKey,
-                                            const RSAPadding &padding,
-                                            const std::string message)
+// ECC
+void SignatureUtils::setUpContext(EVP_PKEY_CTX *ctx,
+                                  const OperationType &operation)
+{
+    if (operation == OperationType::Sign) {
```
Comment (pmvfonseca) [inline src/signature.cpp]: A conditional/ternary operator would look cooler here :smirk:

File: src/signature.cpp [hunk 19]
```diff
@@ -26,9 +26,168 @@ namespace mococrw {
...
+    }
+    _EVP_PKEY_CTX_set_signature_md(ctx, _getMDPtrFromDigestType(openssl::DigestTypes::SHA1));
+}
+
+// RSA
+void SignatureUtils::setUpContext(EVP_PKEY_CTX *ctx,
+                                  const RSAPadding &padding,
+                                  const OperationType &operation)
+{
+    if (operation == OperationType::Sign) {
```
Comment (pmvfonseca) [inline src/signature.cpp]: See above

File: src/signature.cpp [hunk 20]
```diff
@@ -26,9 +26,168 @@ namespace mococrw {
...
+}
+
+/*********************************** ECC specialization  ***********************************/
+std::vector<uint8_t> SignatureUtils::ECC::create(AsymmetricPrivateKey &privateKey,
+                                                 const std::string &message)
+{
+    std::vector<uint8_t> messageDigest;
+
+    try {
+        messageDigest = digestMessage(message,  DigestTypes::SHA1);
```
Comment (pmvfonseca) [inline src/signature.cpp]: Why is only SHA1 supported at the moment? It is considered to be insecure. Shouldn't the user have the option to choose the digest algorithm?

Furthermore, if SHA1 remains as the only supported algorithm then the documentation should emphasize this.

File: src/signature.cpp [hunk 21]
```diff
@@ -26,9 +26,168 @@ namespace mococrw {
...
+
+void SignatureUtils::ECC::verify(AsymmetricPublicKey &publicKey,
+                                 const std::vector<uint8_t> &signature,
+                                 const std::string &message)
+{
+    std::vector<uint8_t> messageDigest;
+
+    try {
+        messageDigest = digestMessage(message,  DigestTypes::SHA1);
+
```
Comment (pmvfonseca) [inline src/signature.cpp]: Remove the extra line :)

File: tests/unit/test_signature.cpp [hunk 22]
```diff
@@ -324,97 +385,174 @@ TEST_F(SignatureTest, testSuccessfulVerificationWithValidPublicKey)
...
+ * @brief Unsucessful verification due to an invalid ECC public key.
+ */
+TEST_F(SignatureTest, testUnsuccessfulEccVerificationWithWrongPublicKey)
+{
+    auto publicKey = mococrw::AsymmetricKeypair::readPublicKeyFromPEM(_invalidEccPublicKey);
+
+    ASSERT_EQ(publicKey.getType(), AsymmetricKey::KeyTypes::ECC);
+    ASSERT_THROW(SignatureUtils::ECC::verify(publicKey, _validEccSignatureSHA1, _testMessage),
+                 MoCOCrWException);
+}
```
Comment (goncalogouveiacriticaltechworkscom) [inline tests/unit/test_signature.cpp]: Missing new line

File: tests/unit/test_signature.cpp [hunk 23]
```diff
@@ -35,27 +36,41 @@ class SignatureTest : public ::testing::Test
...
-    static const std::string _validPublicKey;
-    static const std::string _validPrivateKey;
-
-    // The signatures of the hashed test message with the previously declared private key
+    /* The SHA1 digest of the test message. */
+    static const std::vector<uint8_t> _testMessageDigestSHA1;
+
+    /* Valid keys belonging to the same key pair. */
+    static const std::string _validRsaPublicKey;
+    static const std::string _validEccPublicKey;
```
Comment (pmvfonseca) [inline tests/unit/test_signature.cpp:59]: Maybe we should add a small comment with the information of what curve was chosen to generate the keys.

File: tests/unit/test_signature.cpp [hunk 24]
```diff
@@ -35,27 +36,41 @@ class SignatureTest : public ::testing::Test
     void SetUp() override;
 
 protected:
-    std::unique_ptr<AsymmetricKeypair> _keyPair;
+    std::unique_ptr<AsymmetricKeypair> _keyPairRsa;
```
Comment (pmvfonseca) [inline tests/unit/test_signature.cpp:39]: Small nitpick that applies to the whole file: I think RSA and ECC should always be fully capitalized.

**Review by pmvfonseca — CHANGES_REQUESTED** [review]
Nice job! I like the new class structure, it's more elegant than before.

File: src/mococrw/signature.h (hunk 13 above)
Comment (joaoferreira-ctw) [reply src/mococrw/signature.h]: Agreed. Changed as suggested.

File: src/mococrw/signature.h (hunk 16 above)
Comment (joaoferreira-ctw) [reply src/mococrw/signature.h]: Yes it does...

File: src/mococrw/signature.h (hunk 17 above)
Comment (joaoferreira-ctw) [reply src/mococrw/signature.h]: Agree. Added info to the previous method.

File: src/mococrw/signature.h (hunk 14 above)
Comment (joaoferreira-ctw) [reply src/mococrw/signature.h]: Corrected.

File: src/mococrw/signature.h (hunk 15 above)
Comment (joaoferreira-ctw) [reply src/mococrw/signature.h]: Corrected.

**Review by goncalogouveiacriticaltechworkscom — CHANGES_REQUESTED** [review]
Nice work!
Just check my review comment.

File: src/signature.cpp (hunk 18 above)
Comment (joaoferreira-ctw) [reply src/signature.cpp]: Looking cool indeed!

File: src/signature.cpp (hunk 19 above)
Comment (joaoferreira-ctw) [reply src/signature.cpp]: Corrected.

File: src/signature.cpp (hunk 21 above)
Comment (joaoferreira-ctw) [reply src/signature.cpp]: Removed.

File: tests/unit/test_signature.cpp (hunk 24 above)
Comment (goncalogouveiacriticaltechworkscom) [reply tests/unit/test_signature.cpp:39]: I think code conventions state to just capitalize the first letter.

File: src/signature.cpp (hunk 18 above)
Comment (goncalogouveiacriticaltechworkscom) [reply src/signature.cpp]: I prefer the way it is implemented :)

File: src/signature.cpp (hunk 20 above)
Comment (joaoferreira-ctw) [reply src/signature.cpp]: OpenSSL documentation claims to support SHA1 exclusively, but SHA256 and SHA512 also work as confirmed by unit tests. Added support for SHA256 and SHA512.

File: tests/unit/test_signature.cpp (hunk 24 above)
Comment (joaoferreira-ctw) [reply tests/unit/test_signature.cpp:39]: Agree. Rejected first comment.

File: tests/unit/test_signature.cpp (hunk 23 above)
Comment (joaoferreira-ctw) [reply tests/unit/test_signature.cpp:59]: Agreed. Added.

File: tests/unit/test_signature.cpp (hunk 22 above)
Comment (joaoferreira-ctw) [reply tests/unit/test_signature.cpp]: Alt Text

**Review by goncalogouveiacriticaltechworkscom — APPROVED** [review]
Looks good to me.

**Review by pmvfonseca — APPROVED** [review]

**Pull request merged**